File: frontend/src/types/patient.ts
export interface MedicalConditionEntry {
  condition: string;
  details: string | null;
  currently_in_treatment: boolean | null;
}

export interface Patient {
  id: string;
  hospital_id: string;
  patient_reference_number: string;
  title?: string | null;
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: string;
  blood_group: string | null;
  phone_country_code: string;
  phone_number: string;
  email: string | null;
  address_line_1: string;
  address_line_2: string | null;
  city: string | null;
  state: string | null;
  pin_code: string | null;
  country: string | null;
  age_years: number | null;
  age_months: number | null;
  marital_status: string | null;
  emergency_contact_name: string | null;
  emergency_contact_phone: string | null;
  emergency_contact_country_code: string | null;
  emergency_contact_relation: string | null;
  photo_url: string | null;
  known_allergies: string | null;
  chronic_conditions: string | null;
  // Fixed "Condition / History" checklist (Prescription Builder, below
  // Prescription History) — distinct from the free-text chronic_conditions
  // above; one entry per condition in MEDICAL_CONDITIONS_CHECKLIST.
  medical_conditions: MedicalConditionEntry[] | null;
  // Patient History block (BRD v1.1 §2) — eye-hospital feature pack only
  reason_for_visit: string | null;
  symptoms: string[] | null;
  blood_sugar_value: number | null;
  blood_sugar_unit: string | null;
  // Verification (BRD_OP_1 §3.2) — checkmark requires BOTH true, see
  // utils/patientVerification.ts::isPatientVerified.
  is_email_verified: boolean;
  email_verified_at: string | null;
  is_phone_verified: boolean;
  phone_verified_at: string | null;
  is_verified: boolean;
  is_deleted: boolean;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface PatientListItem {
  id: string;
  patient_reference_number: string;
  first_name: string;
  last_name: string;
  full_name: string;
  date_of_birth: string | null;
  age_years: number | null;
  age_months: number | null;
  gender: string;
  phone_country_code: string;
  phone_number: string;
  email: string | null;
  city: string | null;
  blood_group: string | null;
  known_allergies: string | null;
  chronic_conditions: string | null;
  is_email_verified: boolean;
  is_phone_verified: boolean;
  is_verified: boolean;
  created_at: string;
  updated_at: string;
}

export interface PatientLastVisit {
  last_visit_date: string | null;
}

export interface PaginatedResponse<T> {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  data: T[];
}

export interface PatientCreateData {
  title?: string;
  first_name: string;
  last_name: string;
  // Optional to match backend/app/schemas/patient.py's PatientBase (both are
  // Optional server-side) — Register.tsx's own form still always supplies
  // both since its UI requires them; only the bulk-upload template (which
  // must tolerate a blank cell without failing the whole row) relies on the
  // optionality here.
  date_of_birth?: string;
  gender: string;
  blood_group?: string;
  phone_country_code: string;
  phone_number: string;
  email?: string;
  address_line_1?: string;
  address_line_2?: string;
  city?: string;
  state?: string;
  pin_code?: string;
  country?: string;
  age_years?: number;
  age_months?: number;
  marital_status?: string;
  emergency_contact_name?: string;
  emergency_contact_phone?: string;
  emergency_contact_country_code?: string;
  emergency_contact_relation?: string;
  reason_for_visit?: string;
  symptoms?: string[];
  blood_sugar_value?: number;
  blood_sugar_unit?: string;
}

File: backend/app/models/patient.py
"""
Patient model — matches new hms_db UUID schema.
"""
import uuid
from sqlalchemy import (
    Column, String, Boolean, DateTime, Date, Integer, Text, ForeignKey, UniqueConstraint, Numeric
)
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..database import Base


class Patient(Base):
    __tablename__ = "patients"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    hospital_id = Column(UUID(as_uuid=True), ForeignKey("hospitals.id"), nullable=False)
    patient_reference_number = Column(String(12), nullable=False, index=True)
    title = Column(String(10))
    first_name = Column(String(100), nullable=False)
    last_name = Column(String(100), nullable=False)
    date_of_birth = Column(Date)
    age_years = Column(Integer)
    age_months = Column(Integer)
    gender = Column(String(20), nullable=False)
    blood_group = Column(String(5))
    marital_status = Column(String(20))
    phone_country_code = Column(String(5), nullable=False, default="+1")
    phone_number = Column(String(15), nullable=False)
    secondary_phone = Column(String(20))
    email = Column(String(255))
    national_id_type = Column(String(30))
    national_id_number = Column(String(50))
    address_line_1 = Column(String(255))
    address_line_2 = Column(String(255))
    city = Column(String(100))
    state_province = Column(String(100))
    postal_code = Column(String(20))
    country = Column(String(100), default="India")
    photo_url = Column(String(500))
    emergency_contact_name = Column(String(200))
    emergency_contact_phone = Column(String(20))
    emergency_contact_country_code = Column(String(5), default="+91")
    emergency_contact_relation = Column(String(50))
    known_allergies = Column(Text)
    chronic_conditions = Column(Text)
    notes = Column(Text)
    # Patient History Block fields (BRD v1.1)
    reason_for_visit = Column(Text)
    symptoms = Column(JSONB)  # Array of symptom strings (multi-select + custom entries)
    blood_sugar_value = Column(Numeric(10, 2))
    blood_sugar_unit = Column(String(10))  # 'mg/dL' or 'mmol/L'
    preferred_language = Column(String(10), default="en")
    # Verification (BRD_OP_1 §3.2) — checkmark requires BOTH true, see
    # schemas.patient.PatientListItem.is_verified. Phone/OTP is scaffolded
    # (columns only) but never set True until a future SMS-integration pass.
    is_email_verified = Column(Boolean, default=False)
    email_verified_at = Column(DateTime(timezone=True))
    is_phone_verified = Column(Boolean, default=False)
    phone_verified_at = Column(DateTime(timezone=True))
    is_active = Column(Boolean, default=True)
    registered_at = Column(DateTime(timezone=True), server_default=func.now())
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    updated_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    is_deleted = Column(Boolean, default=False)
    deleted_at = Column(DateTime(timezone=True))

    # Relationships
    hospital = relationship("Hospital", foreign_keys=[hospital_id])

    __table_args__ = (
        UniqueConstraint("hospital_id", "patient_reference_number", name="uq_patient_prn_hospital"),
    )

    @property
    def full_name(self) -> str:
        parts = []
        if self.title:
            parts.append(self.title)
        parts.append(self.first_name)
        parts.append(self.last_name)
        return " ".join(parts)


# Single-use hashed email verification token, mirroring
# backend/app/models/user.py::PasswordResetToken. Phone/OTP has no
# equivalent table yet — that flow is out of scope until an SMS provider
# is integrated (see BRD_OP_1_Development_Plan.md Phase 3d).
class PatientEmailVerificationToken(Base):
    __tablename__ = "patient_email_verification_tokens"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("patients.id", ondelete="CASCADE"), nullable=False)
    token_hash = Column(String(64), unique=True, nullable=False)
    code_hash = Column(String(64))
    attempts_count = Column(Integer, nullable=False, default=0)
    max_attempts = Column(Integer, nullable=False, default=5)
    expires_at = Column(DateTime(timezone=True), nullable=False)
    used_at = Column(DateTime(timezone=True))
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    patient = relationship("Patient")

